### ISSUE TITLE
switch_calling_convention() in CodeGen_C

### ISSUE THREAD

**Issue by jingpu** [halide/Halide#1352]
https://github.com/halide/Halide/blob/b87df1163b406ed7e5c22941592bf185ec883dc9/src/CodeGen_C.cpp#L338-L347
I wonder if the logic here is reversed. It makes more sense to me to emit `extern "C"` before C function prototype in C++ code. Therefore, the function would be written like this:
```
    void switch_calling_convention(bool c_plus_plus) {
      if (in_c_plus_plus != c_plus_plus) {
          if (in_c_plus_plus) {
              stream << "extern \"C\" {\n";
          } else {
              stream << "}\n";
          }
          in_c_plus_plus = c_plus_plus;
      }
    }
```

**Comment by abadams** [comment]
It also reads the wrong way around to me. @zvookin  ?

**Comment by zvookin** [comment]
I think something is more wrong than the two sides being switched. (E.g. Orderings of C and C++ calls, it doesn't look guaranteed that the scopes introduced will be properly closed.) Likely the right answer is to partition all the calls into C and C++ and have all of each emitted in their own block. I'll fix it.

**Comment by zvookin** [comment]
Closed via commit ac9409216f7a2828d3e79690c3a996002bd2f7e9.
